Implemented Floor Again

parth121101 opened the pull request:
Fix #3330

HarshitaKalani reviewed (commented):
You can also add checks for kind of result obtained with:
kind(floor(x, 4)) == 4
kind(floor(x, 8)) == 8

Rest LGTM!

certik commented on src/libasr/pass/intrinsic_function_registry.h:
```diff
@@ -2218,18 +2221,28 @@ namespace Floor {
+        *       r = int(x)
+        *   } else {
+        *       r = int(x) - 1
+        *   }
```
The runtime implementation is slightly different to the compile time implementation. I would keep both exactly the same.

My second question is: is this the correct implementation? I don't like the double if statement. Is it possible to write it using just a single if statement?

HarshitaKalani replied:
* r = floor(x)
        * if ( x <= 0.00 && x != int(x) ) then
        *       r = int(x) - 1
        * else 
        *       r = int(x)
        * end if
        
This would be one simpler approach.

parth121101 replied:
I would try to implement this function using the above approach. Thanks @HarshitaKalani

certik replied:
That's better. It would be great to get rid of the double condition `x <= 0.00 && x != int(x)` somehow. How do other codes implement `floor`?

But until then, let's implement the above, it's reasonably simple. Maybe something like this:

```
    * r = int(x)
    * if ( x <= 0.00 && x != r ) then
    *       r = r - 1
    * end if
```

parth121101 commented:
@certik , @HarshitaKalani I have implemented floor again using the above approach. Apart from that I have added the tests to check the kind of the result. PR for the same : #3368 . Please review that PR

certik commented:
This was closed in favor of #3368.
